edb/edgeql/compiler/stmt.py

```diff
@@ -512,30 +511,50 @@ def compile_DescribeStmt(
                             itemclass)
                     )
 
-                if (itemclass is None or
-                        itemclass is qltypes.SchemaObjectClass.FUNCTION):
-
-                    try:
-                        funcs: Tuple[s_func.Function, ...] = (
-                            ictx.env.schema.get_functions(
-                                name,
-                                module_aliases=ictx.modaliases)
-                        )
-                    except errors.InvalidReferenceError:
-                        pass
-                    else:
-                        for func in funcs:
-                            items.append(func.get_name(ictx.env.schema))
-                        found = True
-
-                if not found:
-                    obj = schemactx.get_schema_object(
-                        objref,
-                        item_type=itemtype,
-                        ctx=ictx,
-                    )
-
-                    items.append(obj.get_name(ictx.env.schema))
+                last_exc = None
+                # Search in the current namespace AND in std.
+                for aliases in [ictx.modaliases, {}]:
+                    # Use the specific modaliases instead of the
+                    # context ones.
+                    ictx.modaliases = cast(Dict[Optional[str], str], aliases)
```

Comment:
So you've overridden `ictx.modaliases` to `{}` in this call, so the following operations can potentially be broken. Instead of this monkey patching I'd add a `modaliases` param to `get_schema_object` and pass `aliases` explicitly (and `ictx.env.schema.get_functions` already has a `module_aliases` arg for that exact purpose).

Reply:
Why is this hack necessary at all?  Unqualified names are always searched in `std` regardless of `modaliases`.

Reply:
> So you've overridden `ictx.modaliases` to `{}` in this call, so the following operations can potentially be broken. 

What operations would be broken? This is a `DESCRIBE`, so we have no operation or expression there, just a name that we need to match. The `modaliases` just specify which is the default module.

> Instead of this monkey patching I'd add a `modaliases` param to `get_schema_object` and pass `aliases` explicitly (and `ictx.env.schema.get_functions` already has a `module_aliases` arg for that exact purpose).

I'm patching the `ictx` because it is used in both the `get_functions` and `get_schema_object` and the latter takes the whole context from which it extracts `modaliases`.

Now that I look at it closer I think I can use `module_aliases` and `module` parameters of those two calls, respectively. This way I can avoid patching the context.

Reply:
> Why is this hack necessary at all? Unqualified names are always searched in `std` regardless of `modaliases`.

The motivation for the hack is to avoid a user-created object/function in default module (from `ictx.modalises`) to mask a `std` built-in with the same name.

Look at the test: it has a user-defined function `all` and a user-defined scalar `all`. If I don't check `std` separately the user-defined function would mask the `std::all` which is not correct for `DESCRIBE all`. Same would happen if I defined a `default::Object`, it would mask the `std::Object` (I'll add that to the test as well). This happens because if the object is found in a default module, the search doesn't proceed to `std`.

Reply:
Why is masking suddenly an issue in the case of `DESCRIBE`?  Yes, you masked a `std` object, and `DESCRIBE` should abide by the common name resolution rules and not invent its own thing. If you want do describe a masked schema object, qualify it

Reply:
> What operations would be broken? This is a DESCRIBE, so we have no operation or expression there, just a name that we need to match. The modaliases just specify which is the default module.

IDK, the code can be refactored later. Violating separation principles by monkey-patching things isn't best practice, period.

> Why is masking suddenly an issue in the case of DESCRIBE? Yes, you masked a std object, and DESCRIBE should abide by the common name resolution rules and not invent its own thing. 

+1.

Reply:
I type `\d all` and the standard library function is missing, because I happen to be in a module that defines it's own `all`. This is bad because the signatures of the `std::all` and the `default::all` could be completely incompatible and thus in an actual query never confused, acting like polymorphic versions.

The argument for harm of masking types is not so straightforward, but still, it's oddly misleading that just because I happen to be in a particular module (via `SET MODULE foo`) I should suddenly not see built-ins.

Reply:
> This is bad because the signatures of the std::all and the default::all could be completely incompatible and thus in an actual query never confused, acting like polymorphic versions.

It's equally confusing to show the std version, as if you type a query with `all()` in the same repl window the std version wouldn't be called anyways. The `\d` command must show what you would actually execute.

That said, `\d` can potentially show other things that are shadowed in a comment, like 

```
db> \d all
function test::all(blah) {
  ...
}

# The test:all() function shadows std::all() function because
# the "test" module is set as default module for this session.
```

Reply:
> If you want do describe a masked schema object, qualify it

How would I know it's masked? This is one way to discover this by using `\d`/`DESCRIBE`. Presumably I use `\d`/`DESCRIBE` precisely because I don't have a perfect recollection about the thing I'm asking about. It's more informative to show builtins even if they are masked, perhaps especially if they are masked to make this fact more obvious.

Reply:
With functions it's weird to not show the polymorphic matches if they happen to be in `std`.

Consider this:
```
CREATE FUNCTION all(a: int64) -> bool USING (a = count(Object));
```
This just makes `all` a polymorphic function, no? I should expect both versions to work... which currently they don't:
```
db0> SELECT all(3);
{false}
db0> SELECT all(true);
error: could not find a function variant all
  ┌─ query:1:8
  │
1 │ SELECT all(true);
  │        ^^^^^^^^^^ error

db0> SELECT std::all(true);
{true}
```
This actually seems like a bug, not a feature. Perhaps I need to update the original ticket to include this issue.

Reply:
> This just makes all a polymorphic function, no?

No, we don't allow cross-module polymorphism.

Reply:
> No, we don't allow cross-module polymorphism.

Is there a reason as to why not? I mean, is it a hard limitation? If it's not a hard limitation, then it's a bit of a foot gun - I add a utility function to my default module, thinking I'll extend, say, `all` to my own [object] type or whatnot. Suddenly my other queries using the built-in `all` on booleans stop working. In shapes, computables, or just in my EdgeQL code. Even though there's technically no ambiguity w.r.t. the function signature.

It's very counter-intuitive that defining this, will break perfectly reasonable queries that use the built-in `len`:
```
module default {
    type Foo {
        multi property vals -> str;
        ...
    }
    function len(a: Foo) -> int64 using (SELECT sum(len(a.vals)));
}
```

Reply:
> Is there a reason as to why not? 

Because it would be a bit crazy and hard to reason about.

> I add a utility function to my default module, thinking I'll extend, say, all to my own [object] type or whatnot.

Sure we can come up with examples rationalizing this. But it would still be a confusing feature in many contexts.

> It's very counter-intuitive that defining this, will break perfectly reasonable queries that use the built-in len:

It would be very counter-intuitive to me (no kidding) if it worked. If you define a `len()` function in your Python script it's not surprising it completely shadows the built-in. Or take C++ that supports function overloading -- it would be weird if defining a "foo()" function in one namespace means the compiler would now "try" "foo()" from other namespaces too.

IMO enabling this kind of cross-module polymorphism is unprecedented and would be confusing to many users including myself.

Reply:
But this isn't any more "cross-module" than the usual "std is always accessible". Why defining a function with a definitely *distict* signature makes `std` unreachable?

`len` and `all` might be odd cases, but this makes it really hard defining a `to_str` for my own scalar or object type without screwing myself over big-time.

My reasoning is: I called `len('as')` so I'm looking for a `default::len(str)` function, but I find `default::len(Foo)`, which is obviously not a match, soooo I keep looking in `std`, when match not found. I mean, they are completely different functions with distinct signatures, why would they clash?

Python is a bad example, because being duck-typed it handles polymorphism differently, by asking you to hide it inside the function body using `isinstance` and such. On the other hand strictly-typed languages are usually quite happy to distinguish signatures based on parameter types (and order). Basically, it's like whatever namespace you're in, you also import all of `std` directly into it and *then* define more things.

Reply:
> But this isn't any more "cross-module" than the usual "std is always accessible". Why defining a function with a definitely distict signature makes std unreachable?

Because our current reference resolution is based on names, and not other things like signatures. You're proposing to implement a new resolution algorithm to be able to magically overload the few functions we have in `std::`. I'm standing by what I said earlier -- I find the proposed behavior confusing and the motivation to implement it weak.

> len and all might be odd cases, but this makes it really hard defining a to_str for my own scalar or object type without screwing myself over big-time.

Let's say you have more modules than just "default" -- because why would we add a feature that only works when there's one module. In this case let's say your scalar is defined in module "a":

```
module a {
   type myscalar {...}
   function to_str(myscalar) ...
}
```

and now module "b":

```
module b {
   type MyType {
      property foo -> myscalar;
      property foo_name := to_str(.foo)
   }
}
```

Thing is `foo_name` would call `std::to_str`. You can fix it to call `a::to_str()` but that would go against your idea of magically overloading `std::to_str()`.

You can scale your example to queries. Essentially, the idea of overloading std module functions in your custom schema would only work reliably in trivial schemas with one module.  Or we need to add some other mechanisms to be able to achieve your goal.

Ultimately, there's no chance we would implement such an elaborate departure from the current semantics before 1.0.

Reply:
IMO the only way of implementing what you want is to allow to directly overload `std::to_str(MyType)` or something like that. This can be a backwards compatible post 1.0 feature.

Reply:
C++ is closer to our case, but it's not the same. There you can use `using namespace foo` or `using foo::bar`, but there's no implicit `using namespace std`. However, having said that if you import 2 namespaces with functions of identical name and distinct signatures, C++ is happy enough to resolve them.

http://cpp.sh/3sge2

Basically, to the extent that C++ allows function polymorphism it's happy enough to resolve them even if you import them by 2 consecutive `using namespace ...` directives. Well, at least "a C++" allows it, I'm quite a bit rusty in it at this point so I'm not sure how different versions handle it.

So to complete this analogy, I'd expect that EdgeQL works similar to having: 
```c++
using namespace std;
using namespace mymodule; // or whatever is the "current" module
...
```
Although, this analogy is not perfect in case of an exact name/signature match, which should result in `mymodule` masking `std`.

P.S. None of this changes how explicit module names get resolved, obviously those things can only be looked for in that specific module, as the name suggests. This is only for unqualified names.
P.P.S. Our own algorithm obviously has the capacity to find the functions in the "default" module (whatever that may be) or in `std` if none are found. I'm guessing that it's largely a matter of removing an `if` before looking into `std` and also validating that the stuff in `std` has a different signature... I assume we handle those match candidates based on signature already, so that doesn't sound like overly complex code change. Is there perhaps some additional issue in underlying SQL? Polymorphic resolution from inside function body, etc.?

Reply:
> Basically, to the extent that C++ allows function polymorphism it's happy enough to resolve them even if you import them by 2 consecutive using namespace ... directives. 

But it does that when you **explicitly** "use" namespaces, not implicitly.  I.e. we don't have syntax like

```
WITH MODULE a, MODULE b
SELECT to_str(...)
```

Exhibit (A).

> P.S. None of this changes how explicit module names get resolved, obviously those things can only be looked for in that specific module, as the name suggests. This is only for unqualified names.

That was clear.

> I'm guessing that it's largely a matter of removing an if before looking into std and also validating that the stuff in std has a different signature...

Please stop twisting the topic. We are not discussing a removal of an `if`. We're discussing a semantic change to how functions are resolved. Without an explicit syntax like in (A) you're basically proposing to resolve an unqualified function name by looking it up in all modules in the DB until it's found. IMO this is completely unacceptable, regardless of how simple on the surface it is to implement.

Anyways, I don't think you're listening to my arguments so let me just put my -1 here. 

Please address the comments on this PR so that it works correctly with the current semantics of EdgeQL.

Reply:
I'll rewrite this PR to use comments as a way to notify the user about types and functions being masked. After all, `DESCRIBE` and `\d` are supposed to provide insight into the current schema and masking built-ins is useful insight.

Regarding the behavior of unqualified name resolution, I'll open a separate issue where we can give it some proper consideration.